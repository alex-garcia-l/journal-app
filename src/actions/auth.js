import Swal from 'sweetalert2';

import { types } from '../types/types';
import { finishLoading, startLoading } from './ui';
import { firebase, googleAuthProvider } from '../database/firebase-config';

export const startLogin = (email, password) => {
  return (dispatch) => {
    dispatch(startLoading());
    firebase.auth().signInWithEmailAndPassword(email, password)
      .then(({ user }) => {
        dispatch(login(user.uid, user.displayName));
        dispatch(finishLoading());
      })
      .catch(error => {
        dispatch(finishLoading());
        Swal.fire('Error', error.message, 'error');
      });
  }
}

export const startLoginWithGoogle = () => {
  return (dispatch) => {
    firebase.auth().signInWithPopup(googleAuthProvider)
      .then(({ user }) => {
        dispatch(login(user.uid, user.displayName));
      });
  }
}

export const startUserRegister = (name, email, password) => {
  return (dispatch) => {
    dispatch(startLoading());
    firebase.auth().createUserWithEmailAndPassword(email, password)
      .then(async ({ user }) => {
        await user.updateProfile({ displayName: name });
        dispatch(login(user.uid, user.displayName));
        dispatch(finishLoading());
      })
      .catch(error => {
        dispatch(finishLoading());
        Swal.fire('Error', error.message, 'error');
      });
  }
}

export const startLogout = () => {
  return async (dispatch) => {
    await firebase.auth().signOut();
    dispatch(logout());
  }
}

export const login = (uid, displayName) => ({
  type: types.login,
  payload: {
    uid,
    displayName
  }
})

export const logout = () => ({
  type: types.logout
})
