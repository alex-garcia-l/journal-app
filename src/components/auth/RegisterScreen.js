import React from 'react';
import { Link } from 'react-router-dom';

export const RegisterScreen = () => {
  return (
    <>
      <h3 className="auth__title mb-5">
        <i class="fa-solid fa-user-plus fa-3x mb-5"></i>
        <span>Register</span>
      </h3>
      
      <form>
        <input
          type="text"
          placeholder="Name"
          name="name"
          autoComplete="off"
          className="input"
        />

        <input
          type="email"
          placeholder="Email"
          name="email"
          autoComplete="off"
          className="input"
        />

        <input
          type="password"
          placeholder="Password"
          name="password"
          className="input"
        />

        <input
          type="password"
          placeholder="Confirm password"
          name="password_confirm"
          className="input"
        />

        <button
          type="submit"
          className="btn btn-primary btn-block"
        >
          Register
        </button>

        <div className="auth__create_acount">
          <Link
            to="/auth/login"
          >
            Already registered?
          </Link>
        </div>
      </form>
    </>
  );
};
