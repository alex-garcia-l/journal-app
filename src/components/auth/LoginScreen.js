import { useDispatch, useSelector } from 'react-redux';
import { Link } from 'react-router-dom';
import { startLogin, startLoginWithGoogle } from '../../actions/auth';

import { useForm } from '../../hooks/useForm';

export const LoginScreen = () => {

  const dispatch = useDispatch();
  const { loading } = useSelector(state => state.ui);

  const [valuesForm, handleInputChange] = useForm({
    email: '',
    password: ''
  });

  const { email, password } = valuesForm;

  const handleSubmit = (evt) => {
    evt.preventDefault();
    dispatch(startLogin(email, password));
  }

  const handleGoogleLogin = () => {
    dispatch(startLoginWithGoogle());
  }

  return (
    <div className="animate__animated animate__fadeIn animate__faster">
      <h3 className="auth__title mb-5">
        <i className="fa-solid fa-user-large fa-3x mb-5"></i>
        <span>Login</span>
      </h3>
      <form onSubmit={handleSubmit}>
        <input
          type="email"
          placeholder="Email"
          name="email"
          autoComplete="off"
          className="input"
          value={email}
          onChange={handleInputChange}
        />

        <input
          type="password"
          placeholder="Password"
          name="password"
          className="input"
          value={password}
          onChange={handleInputChange}
        />
        <button
          type="submit"
          className="btn btn-primary btn-block"
          disabled={loading}
        >
          Login
        </button>

        <div className="auth__social_networks">
          <p className="auth__paragraph">
            Login with social network
          </p>
          <div className="google-btn" onClick={handleGoogleLogin}>
            <div className="google-icon-wrapper">
              <img className="google-icon" src="https://upload.wikimedia.org/wikipedia/commons/5/53/Google_%22G%22_Logo.svg" alt="google button" />
            </div>
            <p className="btn-text">
              <b>Sign in with google</b>
            </p>
          </div>
        </div>

        <div className="auth__create_acount">
          <Link
            to="/auth/register"
          >
            Create account
          </Link>
        </div>
      </form>
    </div>
  );
};
