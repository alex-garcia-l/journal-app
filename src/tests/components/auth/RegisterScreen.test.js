/**
 * @jest-environment node
 */

import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { mount } from 'enzyme';
import { Provider } from 'react-redux';
import { MemoryRouter } from 'react-router-dom';

import { RegisterScreen } from '../../../components/auth/RegisterScreen';
import { types } from '../../../types/types';

import 'jsdom-global/register';

const middlewares = [thunk];
const mockStore = configureStore(middlewares);

const initialState = {
  auth: {},
  ui: {
    loading: false,
    msgError: null
  }
};

let store = mockStore(initialState);
let wrapper = mount(
  <Provider store={store}>
    <MemoryRouter >
      <RegisterScreen />
    </MemoryRouter>
  </Provider >
);

describe('Pruebas para <RegisterScreen />', () => {

  beforeEach(() => {
    store = mockStore(initialState);

    wrapper = mount(
      <Provider store={store}>
        <MemoryRouter >
          <RegisterScreen />
        </MemoryRouter>
      </Provider >
    );
  });

  test('debe renderizar correctamente', () => {
    expect(wrapper).toMatchSnapshot();
  });

  test('debe marcar error al registrarse', () => {

    wrapper.find('form').simulate('submit', {
      preventDefault: () => { }
    });

    const actions = store.getActions();

    expect(actions[0]).toEqual({
      type: types.uiSetError,
      payload: 'Name is required.'
    });

  });

  test('debe mostrar el div del error', () => {

    const initialState = {
      auth: {},
      ui: {
        loading: false,
        msgError: 'Name is required.'
      }
    };

    let store = mockStore(initialState);
    let wrapper = mount(
      <Provider store={store}>
        <MemoryRouter >
          <RegisterScreen />
        </MemoryRouter>
      </Provider >
    );

    expect(wrapper.find('.auth__alert-error').exists()).toBe(true);
    expect(wrapper.find('.auth__alert-error').text().trim()).toBe(initialState.ui.msgError);

  });

});
