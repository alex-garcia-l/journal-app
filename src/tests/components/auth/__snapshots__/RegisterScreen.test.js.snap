// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas para <RegisterScreen /> debe renderizar correctamente 1`] = `
<div
  className="animate__animated animate__fadeIn animate__faster"
>
  <h3
    className="auth__title mb-5"
  >
    <i
      className="fa-solid fa-user-plus fa-3x mb-5"
    />
    <span>
      Register
    </span>
  </h3>
  <form
    onSubmit={[Function]}
  >
    <input
      autoComplete="off"
      className="input"
      name="name"
      onChange={[Function]}
      placeholder="Name"
      type="text"
      value=""
    />
    <input
      autoComplete="off"
      className="input"
      name="email"
      onChange={[Function]}
      placeholder="Email"
      type="email"
      value=""
    />
    <input
      className="input"
      name="password"
      onChange={[Function]}
      placeholder="Password"
      type="password"
      value=""
    />
    <input
      className="input"
      name="passwordConfirm"
      onChange={[Function]}
      placeholder="Confirm password"
      type="password"
      value=""
    />
    <button
      className="btn btn-primary btn-block"
      disabled={false}
      type="submit"
    >
      Register
    </button>
    <div
      className="auth__create_acount"
    >
      <a
        href="/auth/login"
        onClick={[Function]}
      >
        Already registered?
      </a>
    </div>
  </form>
</div>
`;
